jsonata.js

```diff
@@ -2219,6 +2224,20 @@ var jsonata = (function() {
                         value: key
                     };
                 }
+                // check that this key isn't associated with a different expression
+                if(owners.hasOwnProperty(key)) {
+                    if(owners[key] !== pairIndex) {
+                        // this key has been generated by another expression in this group
+                        throw {
+                            code: "D1009",
+                            stack: (new Error()).stack,
+                            position: expr.position,
+                            value: key
+                        };
+                    }
+                } else {
+                    owners[key] = pairIndex;
+                }
                 var entry = {data: item, expr: pair[1]};
```

Comment:
Note that instead of storing `expr` here, you could simply store the `pairIndex` and then lookup the expression later as `expr.lhs[pairIndex][1]`.  The point being that what you are trying to do here is avoid having to compare expressions for equality (which isn't feasible) so you are instead storing the index of the pair whose value expression you are using.  But you could use that `pairIndex` everywhere and not store the expression.  Then, you'd only need to see if the pairIndex matched when you attempted to append a new item to `data`.

Said another way, don't store `expr` in the `entry`, store the `pairIndex`.  Then if you find a key already exists, you won't even need the `owners` object, just check to make sure that you have the same `pairIndex` as you did when the entry was created.

Does that make sense?

Reply:
Alternatively, you can store `expr` **and** `pairIndex`.  Same effect, you don't need `owners` at all.

Reply:
That's true.  I'll make that change. Thanks!